components/org.wso2.carbon.identity.api.user.recovery/org.wso2.carbon.identity.rest.api.user.recovery.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/recovery/v1/impl/core/PasswordRecoveryService.java

```diff
@@ -0,0 +1,466 @@
+                            RecoveryUtil.buildPropertiesMap(initRequest.getProperties()));
+            // If RecoveryChannelInfoDTO is null throw not found error.
+            if (recoveryInformationDTO == null) {
+                if (log.isDebugEnabled()) {
+                    String message = "No recovery information for password recovery request";
+                    log.debug(message);
+                }
+                return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+            }
+            return Response.ok().entity(buildPasswordRecoveryInitResponse(tenantDomain, recoveryInformationDTO))
+                    .build();
+        } catch (IdentityRecoveryClientException e) {
+            throw RecoveryUtil.handleClientExceptions(PasswordRecoveryService.class.getName(), tenantDomain,
+                    IdentityRecoveryConstants.PASSWORD_RECOVERY_SCENARIO, Util.getCorrelation(), e);
+        } catch (IdentityRecoveryException e) {
+            throw RecoveryUtil.buildInternalServerErrorResponse(PasswordRecoveryService.class.getName(),
+                    Constants.SERVER_ERROR, e.getErrorCode(), Util.getCorrelation(), e);
+        } catch (Throwable throwable) {
+            throw RecoveryUtil
+                    .buildInternalServerErrorResponse(PasswordRecoveryService.class.getName(),
+                            Constants.SERVER_ERROR,
+                            IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED_ERROR.getCode(), Util
+                                    .getCorrelation(), throwable);
+        }
+    }
+
+    /**
+     * Send the recovery notifications to the user via user preferred channel given by the channelId param.
+     *
+     * @param recoveryRequest {@link RecoveryRequest} Object which holds the recovery information in the
+     *                        password recovery request
+     * @return Response
+     */
+    public Response recoverPassword(RecoveryRequest recoveryRequest) {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        String recoveryId = recoveryRequest.getRecoveryCode();
+        String channelId = recoveryRequest.getChannelId();
+        try {
+            PasswordRecoverDTO passwordRecoverDTO = UserAccountRecoveryServiceDataHolder.getPasswordRecoveryManager()
+                    .notify(recoveryId, channelId, tenantDomain,
+                            RecoveryUtil.buildPropertiesMap(recoveryRequest.getProperties()));
+            if (passwordRecoverDTO == null) {
+                if (log.isDebugEnabled()) {
+                    String message = String
+                            .format("No password recovery data object for recovery code : %s", recoveryId);
+                    log.debug(message);
+                }
+                return Response.status(Response.Status.OK).build();
```

Comment:
If the code to reach this line without any errors, that means recovery is successful but reponse object is no there. 
Using 404 is wrong since the 404 error status code indicates that the REST API can’t map the client’s URI to a resource but may be available in the future. 

Therefore 500 will be a better option. WDYT?